src/utilities/types.ts

```diff
@@ -11,3 +11,16 @@ export type Picked<T extends PlainObject, Keys extends Array<keyof T>> = {
 export type Falsy = '' | 0 | false | null | undefined;
 
 export type Truthy<T> = Exclude<T, Falsy>;
+
+/* Copied from react-i18next/ts4.1/index.d.ts */
+// Normalize single namespace
+type AppendKeys<K1, K2> = `${K1 & string}.${K2 & string}`;
+type AppendKeys2<K1, K2> = `${K1 & string}.${Exclude<K2, keyof any[]> & string}`;
+type Normalize2<T, K = keyof T> = K extends keyof T
+  ? T[K] extends Record<string, any>
+    ? T[K] extends readonly any[]
+      ? AppendKeys2<K, keyof T[K]> | AppendKeys2<K, Normalize2<T[K]>>
+      : AppendKeys<K, keyof T[K]> | AppendKeys<K, Normalize2<T[K]>>
+    : never
+  : never;
+export type Normalize<T> = keyof T | Normalize2<T>;
```

Comment:
maybe a better naming with comments will help to understand?

Reply:
Improved, PTAL